pytorch_lightning/utilities/auto_restart.py

```diff
@@ -476,6 +430,48 @@ def _capture_metadata_collate(
     return {"data": data, AutoRestartBatchKeys.PL_RESTART_META: metadata}
 
 
+# TODO: Merge this code within stateful DataLoaderIter.
+def _next_data_wrapper(fn, it, dl, num_batches_fetched, data_fetcher) -> Callable:
+    @wraps(fn)
+    def wrapper():
+        nonlocal num_batches_fetched
+        nonlocal it
+        nonlocal dl
+
```

Comment:
These serve no purpose now afaik

```suggestion
```

num_batches_fetched is not used anywhere else outside of the wrapper and it and dl would be modified by reference anyways

Reply:
You need the `nonlocal num_batches_fetched` or it would be unbounded.